### PR TITLE
Add cache to viewdescriptorservice

### PR DESCRIPTION
**Pull request by sbatten**
refs #85164

### CODE REVIEW

**Comment by sandy081** [comment]
@sbatten Not sure why this looks bit complicated. I am guessing the implementation as follows:

- Create a `IViewDescriptorsCollection` for a given container using the location information from `ViewDescriptorsRegistry` and Storage (cache). 
- You primary source for view descriptors should be `ViewDescriptorsRegistry`. Listen to `ViewDescriptorsRegistry` and update view descriptors of `IViewDescriptorsCollection` based on the location information.

Let me know if I am missing anything else here?

**Comment by sandy081** [comment]
@sbatten 

I think we can do a bit of clean up without your changes and after that your changes can become more simpler. for eg: try removing the events `onViewsRegistered` and `onViewsDeregistered` from the service. These events are only used inside the service.

File: src/vs/workbench/browser/parts/views/views.ts [hunk 1]
```diff
@@ -796,41 +859,106 @@ export class ViewDescriptorService extends Disposable implements IViewDescriptor
...
+	private onDidRegisterViewContainer(viewContainer: ViewContainer): void {
+		const disposables = new DisposableStore();
+		const viewDescriptorCollection = disposables.add(new ViewDescriptorCollection(this.contextKeyService));
 
 		this.onDidChangeActiveViews({ added: viewDescriptorCollection.activeViewDescriptors, removed: [] });
 		viewDescriptorCollection.onDidChangeActiveViews(changed => this.onDidChangeActiveViews(changed), this, disposables);
 
 		this.viewDescriptorCollections.set(viewContainer, { viewDescriptorCollection, disposable: disposables });
+
+		// Check if we have any views that were awaiting container registration due to a cached position
```
Comment (sandy081) [inline src/vs/workbench/browser/parts/views/views.ts]: Looks this comment is outdated

File: src/vs/workbench/browser/parts/views/views.ts [hunk 2]
```diff
@@ -796,41 +859,106 @@ export class ViewDescriptorService extends Disposable implements IViewDescriptor
...
+			this.addViews(to, views);
+			this.saveViewPositionsToCache();
 		}
 	}
 
-	private onDidRegisterViewContainer(viewContainer: ViewContainer): void {
-		const disposables = new DisposableStore();
-		const viewDescriptorCollection = disposables.add(new ViewDescriptorCollection(this.contextKeyService));
-		viewDescriptorCollection.addViews(this.viewsRegistry.getViews(viewContainer));
+	private getViewContainerById(id: string): ViewContainer | undefined {
```
Comment (sandy081) [inline src/vs/workbench/browser/parts/views/views.ts]: Is this method does not exist on view container registry? If does not, lets please create one there and use it.

File: src/vs/workbench/browser/parts/views/views.ts [hunk 3]
```diff
@@ -734,54 +720,131 @@ export class ViewDescriptorService extends Disposable implements IViewDescriptor
...
+		this.registerGroupedViews(regroupedViews);
+	}
+
+	private onDidDeregisterViews(views: IViewDescriptor[], viewContainer: ViewContainer): void {
+		// When views are registered, we need to regroup them based on the cache
+		const regroupedViews = this.regroupViews(viewContainer.id, views);
+		this.deregisterGroupedViews(regroupedViews);
+	}
+
+	private regroupViews(containerId: string, views: IViewDescriptor[], seedMap?: Map<string, IViewDescriptor[]>): Map<string, IViewDescriptor[]> {
```
Comment (sandy081) [inline src/vs/workbench/browser/parts/views/views.ts]: I do not see `seedMap` is passed any where. So this arg is not needed.

**Review by sandy081 — APPROVED** [review]

**Comment by sandy081** [comment]
LGTM. Provided some feedback.

**Comment by sbatten** [comment]
will update and merge thanks

**Pull request merged**